extensions/amp-story/0.1/page-scaling.js

```diff
@@ -0,0 +1,457 @@
+   */
+  measureScalableElsFor(page) {
+    const {width, height} = unscaledClientRect(page);
+    const pageWidth = width;
+    const pageHeight = height;
+    return scalableElements(page).map(el => {
+      const {width, height} = unscaledClientRect(el);
+      return {
+        matrix: this.getTransformMatrix(el),
+        relativeWidth: width / pageWidth,
+        relativeHeight: height / pageHeight,
+      };
+    });
+  }
+
+  /**
+   * Gets an element's transform matrix.
+   * @param {!Element} unusedEl
+   * @return {?Array<number>}
+   */
+  getTransformMatrix(unusedEl) {
+    // Calculating a transform matrix is optional depending on scaling
+    // implementation.
+    return null;
+  }
+
+  /** @private */
+  onViewportResize_() {
+    this.targetDimensions_ = null;
+    this.vsync_.measure(() => {
+      this.measureTargetDimensions_();
+    });
+    this.updatePagesOnResize();
+  }
+
+  /** @protected */
+  updatePagesOnResize() {
+    const pages = toArray(childElementsByTag(this.rootEl_, 'amp-story-page'));
+    pages.forEach(page => {
+      markScalingApplied(page, false);
+      this.scale_(page);
+    });
+  }
+}
+
+
+/** Uses CSS zoom as scaling method. */
+class ZoomScalingService extends PageScalingService {
+  /** @override */
+  scalingStyles(targetDimensions, elementDimensions) {
+    const {width, height, factor} = targetDimensions;
+    const {relativeWidth, relativeHeight} = elementDimensions;
+    return {
+      'width': px(width * relativeWidth),
+      'height': px(height * relativeHeight),
+      'zoom': factor,
+    };
+  }
+}
+
+
+/** Uses combined CSS transform as scaling method. */
+class TransformScalingService extends PageScalingService {
+  /** @override */
+  getTransformMatrix(unusedEl) {
+    // TODO(alanorozco, #12934): Implement.
+    return [1, 0, 0, 1, 0, 0];
+  }
+
+  /** @override */
+  scalingStyles(targetDimensions, elementDimensions) {
+    const {width, height, factor} = targetDimensions;
+    const {relativeWidth, relativeHeight, matrix} = elementDimensions;
+    const initialMatrix = /** @type {!Array<number>} */ (
+        dev().assert(matrix, 'No initial matrix'));
+    const transformedMatrix =
+        scaleTransform(factor, width, height, initialMatrix);
+    return {
+      'width': px(width * relativeWidth),
+      'height': px(height * relativeHeight),
+      'transform': `matrix(${transformedMatrix.join()})`,
+    };
+  }
+}
+
+
+/** Uses CSS zoom and custom CSS properties as scaling method. */
+class CssPropsZoomScalingService extends PageScalingService {
+  /** @override */
+  measureScalableElements(page) {
+    // Circumvents layer cache as layers are only mutated once per page.
+    return this.measureScalableElsFor(page);
+  }
+
+  /** @override */
+  updateRootProps() {
+    const {width, height, factor} = this.targetDimensions_;
+    this.vsync_.mutate(() => {
+      this.rootEl_.style.setProperty('--i-amphtml-story-width', px(width));
```

Comment:
Is the goal that this is user-overridable?  I think the fact that it's `--i-amphtml-*` will make the doc invalid if people try to override this.

Reply:
It's not user-overridable.